### PR TITLE
Race condition in RunCmdWithOutputParser

### PR DESCRIPTION
**Pull request by eranturgeman**
- [x] All [tests](https://github.com/jfrog/gofrog#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
- [x] I labeled this pull request with one of the following: 'breaking change', 'new feature', 'bug', or 'ignore for release'

---

### CODE REVIEW

File: io/cmd.go [hunk 1]
```diff
@@ -94,9 +94,11 @@ func RunCmdWithOutputParser(config CmdConfig, prompt bool, regExpStruct ...*CmdO
 			for _, regExp := range regExpStruct {
 				matched := regExp.RegExp.Match([]byte(line))
 				if matched {
-					regExp.MatchedResults = regExp.RegExp.FindStringSubmatch(line)
-					regExp.Line = line
-					line, err = regExp.ExecFunc(regExp)
+					results := CmdOutputPattern{
+						MatchedResults: regExp.RegExp.FindStringSubmatch(line),
+						Line:           line,
```
Comment (yahavi) [inline io/cmd.go]:
```suggestion
						RegExp:         regExp.RegExp,
						MatchedResults: regExp.RegExp.FindStringSubmatch(line),
						Line:           line,
						ExecFunc:       regExp.ExecFunc,
```

Even better if you could export this logic to a new function and reuse it

File: io/cmd.go [hunk 2]
```diff
@@ -117,8 +119,11 @@ func RunCmdWithOutputParser(config CmdConfig, prompt bool, regExpStruct ...*CmdO
 			for _, regExp := range regExpStruct {
 				matched := regExp.RegExp.Match([]byte(line))
 				if matched {
-					regExp.MatchedResults = regExp.RegExp.FindStringSubmatch(line)
-					regExp.Line = line
+					results := CmdOutputPattern{
+						MatchedResults: regExp.RegExp.FindStringSubmatch(line),
+						Line:           line,
```
Comment (yahavi) [inline io/cmd.go]:
```suggestion
						RegExp:         regExp.RegExp,
						MatchedResults: regExp.RegExp.FindStringSubmatch(line),
						Line:           line,
						ExecFunc:       regExp.ExecFunc,
```

**Review by yahavi — CHANGES_REQUESTED** [review]

**Comment by yahavi** [comment]
* Fix race condition in RunCmdWithOutputParser
* Update Go to 1.20
* Move `Command` struct from build-info-go

Dependency for https://github.com/jfrog/build-info-go/pull/216 & https://github.com/jfrog/jfrog-cli-core/pull/1053

File: io/cmd.go [hunk 3]
```diff
@@ -158,6 +142,34 @@ func RunCmdWithOutputParser(config CmdConfig, prompt bool, regExpStruct ...*CmdO
...
+				ExecFunc:       regExp.ExecFunc,
+			}
+			processedLine, err = regExp.ExecFunc(&results)
+			if err != nil {
+				errChan <- err
+				hasError = true
+				break
+			}
+		}
+	}
```
Comment (attiasas) [inline io/cmd.go]:
```suggestion
		if !regExp.RegExp.MatchString(processedLine) {
			continue
		}
		results := CmdOutputPattern{
			RegExp:         regExp.RegExp,
			MatchedResults: regExp.RegExp.FindStringSubmatch(processedLine),
			Line:           processedLine,
			ExecFunc:       regExp.ExecFunc,
		}
		if processedLine, err = regExp.ExecFunc(&results); err != nil {
			errChan <- err
			hasError = true
			break
		}
	}
```

File: io/cmd_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+var matchAllRegexp = regexp.MustCompile(".*")
+var errParsing = errors.New("parsing error")
+
+func TestRunCmdWithOutputParser(t *testing.T) {
+	config := NewCommand("go", "", []string{"version"})
+	stdout, stderr, exitOk, err := RunCmdWithOutputParser(config, false, &CmdOutputPattern{
+		RegExp:   matchAllRegexp,
+		ExecFunc: func(pattern *CmdOutputPattern) (string, error) { return pattern.Line, nil },
+	})
+	assert.NoError(t, err)
```
Comment (attiasas) [inline io/cmd_test.go]:
```suggestion
	parserCalled := false
	stdout, stderr, exitOk, err := RunCmdWithOutputParser(config, false, &CmdOutputPattern{
		RegExp:   matchAllRegexp,
		ExecFunc: func(pattern *CmdOutputPattern) (string, error) {
			parserCalled = true
			return pattern.Line, nil 
		},
	})
	assert.NoError(t, err)
	assert.True(t, parserCalled)
```
lets also make sure the parser is also called at this process

File: io/cmd_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+	{[]*CmdOutputPattern{{
+		RegExp:   matchAllRegexp,
+		ExecFunc: func(pattern *CmdOutputPattern) (string, error) { return "", errParsing },
+	}}, "hello", "", true},
+}
+
+func TestProcessLine(t *testing.T) {
+	for _, testCase := range processLineCases {
+		t.Run("", func(t *testing.T) {
```
Comment (attiasas) [inline io/cmd_test.go]: lets add names to the tests to have clear logs for them and not:
```
--- PASS: TestRunCmdWithOutputParserError (0.00s)
=== RUN   TestProcessLine
=== RUN   TestProcessLine/#00
=== RUN   TestProcessLine/#01
=== RUN   TestProcessLine/#02
=== RUN   TestProcessLine/#03
=== RUN   TestProcessLine/#04
=== RUN   TestProcessLine/#05
--- PASS: TestProcessLine (0.00s)
    --- PASS: TestProcessLine/#00 (0.00s)
    --- PASS: TestProcessLine/#01 (0.00s)
    --- PASS: TestProcessLine/#02 (0.00s)
    --- PASS: TestProcessLine/#03 (0.00s)
    --- PASS: TestProcessLine/#04 (0.00s)
    --- PASS: TestProcessLine/#05 (0.00s)
```

**Review by attiasas — COMMENTED** [review]
Nice! please consider my comments

**Pull request merged**